### PR TITLE
Add docker image and refactor release pipeline

### PR DESCRIPTION
**Pull request by luhring**
This PR replaces the GoReleaser-based release process with discrete steps as `make` targets.

The main problem addressed here is that we can run release steps on either Linux or macOS, based on what's needed by the step. Before, with GoReleaser doing all steps monolithically, this wasn't possible; and specifically, we could not both **sign our macOS assets** and **build container images**. (Thus #228)

This PR also introduces the concept of a "release follow-up pipeline". Before, we were publishing "draft" releases, but some assets had already been published as final (e.g. the version check file, the homebrew formula update). This PR separates out "follow-up" steps which won't trigger until the draft release has been published.

With these changes in place, this PR also adds a container image build and push mechanism into the release follow-up pipeline.

This PR also tests more artifacts before publishing now... specifically the homebrew formula and the container image.

Other lower-level changes:
- Changes make shell from `sh` to `bash` (to allow for more elegant scripts, e.g. allows `pushd`/`popd`)
- Adds protective shell flags (`-e`, `-o pipefail`)
- Introduces the special `.SILENT:` make target to silence recipe steps more elegantly than prefixing with `@`
- Disambiguates `VERSION` and `VERSION_TAG`. This follows from GoReleaser's separation of the `VERSION` and `TAG` template variables. This distinction matters when used in context (e.g. `VERSION` is `0.12.2`, whereas `VERSION_TAG` is `v0.12.2`).

**This PR does not add documentation** for using the container image. This documentation will be introduced in a future PR, once the new container workflow is confirmed to work, and will accompany other minor docs improvements.

**Note: This PR is likely to require follow-up fixes since it is difficult to test in its entirety with non-real releases.**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -257,31 +358,117 @@ changelog-unreleased: ## show the current changelog that will be produced on the
...
+		"-t $(CONTAINER_IMAGE_TAG_MINOR)" \
+		"-t $(CONTAINER_IMAGE_TAG_PATCH)" \
+		"-t $(CONTAINER_IMAGE_TAG_LATEST)" \
+	) && \
+	DOCKER_BUILDKIT=1 docker build --no-cache $${tags[@]} -f "./Dockerfile" .
+	# Using buildkit due to https://github.com/moby/moby/issues/37965
+
+.PHONY: container-image-test
+.SILENT: container-image-test
+container-image-test:
```
Comment (alfredodeza) [inline Makefile:444]: excellent, thanks for adding a dry-run mode

File: Makefile [hunk 2]
```diff
@@ -257,31 +358,117 @@ changelog-unreleased: ## show the current changelog that will be produced on the
...
+	$(call title,Building and tagging container image for $(BIN))
+
+	tags=( \
+		"-t $(CONTAINER_IMAGE_TAG_MAJOR)" \
+		"-t $(CONTAINER_IMAGE_TAG_MINOR)" \
+		"-t $(CONTAINER_IMAGE_TAG_PATCH)" \
+		"-t $(CONTAINER_IMAGE_TAG_LATEST)" \
+	) && \
+	DOCKER_BUILDKIT=1 docker build --no-cache $${tags[@]} -f "./Dockerfile" .
+	# Using buildkit due to https://github.com/moby/moby/issues/37965
```
Comment (alfredodeza) [inline Makefile:440]: thanks for adding a link explaining the why

File: Makefile [hunk 3]
```diff
@@ -257,31 +358,117 @@ changelog-unreleased: ## show the current changelog that will be produced on the
...
+.PHONY: homebrew-formula-publish
+.SILENT: homebrew-formula-publish
+homebrew-formula-publish:
+	$(call title,Publishing homebrew formula)
+
+	FORMULA_FILE="$$(realpath $(HOMEBREW_FORMULA_FILE))" && \
+	\
+	pushd "$(TEMPDIR)" && \
+		rm -rfv "./homebrew-syft" && \
+		gh repo clone anchore/homebrew-syft && \
```
Comment (alfredodeza) [inline Makefile:411]: where is `gh` coming from?

File: Makefile [hunk 4]
```diff
@@ -257,31 +358,117 @@ changelog-unreleased: ## show the current changelog that will be produced on the
...
+	export DARWIN_AMD64_ASSET_SHA256=$(call sha256,"$(TEMPDIR)/darwin_amd64_asset") && \
+	\
+	export LINUX_AMD64_ASSET_URL=$(call asset_url,"$(TEMPDIR)/release.json","linux_amd64.tar.gz") && \
+	curl -sSL "$${LINUX_AMD64_ASSET_URL}" > "$(TEMPDIR)/linux_amd64_asset" && \
+	export LINUX_AMD64_ASSET_SHA256=$(call sha256,"$(TEMPDIR)/linux_amd64_asset") && \
+	\
+	export VERSION=$(call get_version_from_version_tag,$(VERSION_TAG)) && \
+	\
+	cat "./.homebrew-formula-template.rb" | \
+		envsubst > "$(HOMEBREW_FORMULA_FILE)"
```
Comment (alfredodeza) [inline Makefile:382]: `envsubst` is not available everywhere I think, and some systems require you to install the `gettext` package(s) to make it available. Is this dependency accounted for somewhere?

File: Makefile [hunk 5]
```diff
@@ -94,6 +122,8 @@ bootstrap: ## Download and install all go dependencies (+ prep tooling in the ./
 	[ -f "$(TEMPDIR)/golangci" ] || curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b $(TEMPDIR)/ v1.26.0
 	[ -f "$(TEMPDIR)/bouncer" ] || curl -sSfL https://raw.githubusercontent.com/wagoodman/go-bouncer/master/bouncer.sh | sh -s -- -b $(TEMPDIR)/ v0.2.0
 	[ -f "$(TEMPDIR)/goreleaser" ] || curl -sfL https://install.goreleaser.com/github.com/goreleaser/goreleaser.sh | sh -s -- -b $(TEMPDIR)/ v0.140.0
+	[ -f "$(TEMPDIR)/nfpm" ] || curl -sfL curl -sfL https://install.goreleaser.com/github.com/goreleaser/nfpm.sh | sh -s -- -b $(TEMPDIR)/ v2.2.2
+	[ -f "$(BREW_CMD)" ] || (mkdir -p "$(BREW_DIR)" && curl -L https://github.com/Homebrew/brew/tarball/master | tar -xz --strip 1 -C "$(BREW_DIR)")
```
Comment (alfredodeza) [inline Makefile:126]: I thought I would see `gon` here, is that installed somewhere else?

File: .github/workflows/release.yaml [hunk 6]
```diff
@@ -120,19 +159,60 @@ jobs:
...
+          set -eux
+          assets=()
+          for asset in ./Linux-artifacts/*; do
+            assets+=("-a" "$asset")
+          done
+          for asset in ./macOS-artifacts/*; do
+            assets+=("-a" "$asset")
+          done
+          tag_name="${GITHUB_REF##*/}"
+          hub release edit -m "" "${assets[@]}" "$tag_name"
```
Comment (alfredodeza) [inline .github/workflows/release.yaml:213]: I guess `hub` is just available inside an action right?

Comment (luhring) [reply .github/workflows/release.yaml:213]: Exactly (see here for what's installed natively on GHA runners for [Linux](https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu1804-README.md#cli-tools) and [macOS](https://github.com/actions/virtual-environments/blob/main/images/macos/macos-10.15-Readme.md#utilities))

File: Makefile (hunk 3 above)
Comment (luhring) [reply Makefile:411]: Natively on the runners (see [this comment](https://github.com/anchore/syft/pull/310#discussion_r560967271))

File: Makefile (hunk 4 above)
Comment (luhring) [reply Makefile:382]: This is a great point, it's not accounted for or explicitly checked for — I'll change that

File: test/integration/test-fixtures/image-large-apk-data/Dockerfile [hunk 7]
```diff
@@ -1,2 +1,5 @@
-FROM alpine:latest
-RUN apk add tzdata vim alpine-sdk
+FROM alpine@sha256:d9a7354e3845ea8466bb00b22224d9116b183e594527fb5b6c3d30bc01a20378
```
Comment (alfredodeza) [inline test/integration/test-fixtures/image-large-apk-data/Dockerfile:1]: fese9n89llc61

**Review by alfredodeza — APPROVED** [review]
This looks good. I know a few issues popped up that have been resolved. Feel free to merge when ready

File: Makefile (hunk 4 above)
Comment (wagoodman) [reply Makefile]: Replaced with a templated heredoc instead

**Pull request merged**